book/src/dev/rfcs/0003-state-updates.md

```diff
@@ -0,0 +1,305 @@
+
+Performs contextual validation of the given block, committing it to the state
+if successful. Returns `Response::Added(BlockHeaderHash)` with the hash of
+the newly committed block or an error.
+
+If the parent block is not committed, add the block to an internal queue for
+future processing.
+
+Otherwise, attempt to perform contextual validation checks and the commit the
+given block to the state. The exact list of contextual validation checks will
+be specified in a later RFC. If contextual validation checks succeed, the new
+block is added to one of the in-memory chains. If the resulting chain is
+longer than 100 blocks, the oldest block is now past the reorg limit, so we
+remove it from all in-memory chains and commit it to sled as described below
+in `CommitFinalizedBlock`.
+
+Finally, process any queued children of the newly committed block the same way.
+
+### `Request::CommitFinalizedBlock(Arc<Block>)`
+[request-commit-finalized-block]: #request-finalized-block
+
+Commits a finalized block to the sled state, skipping contextual validation.
+The block's parent must be the current sled tip. This is exposed for use in
+checkpointing, which produces in-order finalized blocks. Returns
+`Response::Added(BlockHeaderHash)` with the hash of the committed block if
+successful.
+
+This should be implemented as a wrapper around a function also called by
+[`Request::CommitBlock`](#request-commit-block), which should:
+
+1. Obtain the highest entry of `hash_by_height` as `(old_height, old_tip)`.
+Check that `block`'s parent hash is `old_tip` and its height is
+`old_height+1`, or panic.  This check is performed as defense-in-depth
+to prevent database corruption, but it is the caller's responsibility to
+commit finalized blocks in order.
+
+2. Insert `(block_hash, block)` into `blocks_by_hash` and
+   `(BE32(height), block_hash)` into `hash_by_height`.
+
+3. Iterate over the enumerated transactions in the block. For each transaction:
+
+   1. Insert `(transaction_hash, block_hash || BE32(tx_index))` to
+   `tx_by_hash`;
+
+   2. For each `TransparentInput::PrevOut { outpoint, .. }` in the
+   transaction's `inputs()`, remove `outpoint` from `utxo_by_output`.
+
+   3. For each `output` in the transaction's `outputs()`, construct the
+   `outpoint` that identifies it, and insert `(outpoint, output)` into
+   `utxo_by_output`.
+
+   4. For each [`JoinSplit`] description in the transaction,
+   insert `(nullifiers[0],())` and `(nullifiers[1],())` into
+   `sprout_nullifiers`.
+
+   5. For each [`Spend`] description in the transaction, insert
+   `(nullifier,())` into `sapling_nullifiers`.
+
+[`JoinSplit`]: https://doc.zebra.zfnd.org/zebra_chain/transaction/struct.JoinSplit.html
+[`Spend`]: https://doc.zebra.zfnd.org/zebra_chain/transaction/struct.Spend.html
+
+    
+These updates can be performed in a batch or without necessarily iterating
+over all transactions, if the data is available by other means; they're
+specified this way for clarity.
+
```

Comment:
Do you also want to describe these cleanup tasks here?
* drop any side chains which are past the reorg limit
* drop any unverified blocks, queued by CommitBlock, which are past the reorg limit

I think it's important to include these steps as part of the design and implementation. Otherwise, we could end up with an implementation which makes them difficult or impossible.

If we want to provide access to side chains for analysis, we can add a RPC call. We can also add an analysis config option, which makes us keep the side chains themselves.

But if we're not doing analysis, it seems risky to keep dangling blocks around - it increases our memory and possibly CPU usage, and increases the size and complexity of our data structures. It also creates an additional attack surface for memory exhaustion. (I think checkpointing and block difficulty might limit these risks on mainnet, but they certainly don't on testnet.)

Reply:
Yes, we should describe these tasks, they're not described at the moment because the current RFC text is focused on the sled structures for data beyond the reorg limit.

I agree that we shouldn't keep dangling blocks around, and I think that the reorg limit is a natural place to draw that boundary, by simply never committing any side-chains to the sled state.